Move some options from transfer list's context menu into "Torrent options" dialog

thalieht opened the pull request:
I can't decide where's the best position for it. Pics of 2 places below (start of PR is the one on top). Let me know if you prefer it in some other position.

![top](https://user-images.githubusercontent.com/6451685/132952634-aea41523-9170-49a0-88d3-39b02f32252c.JPG)![mid](https://user-images.githubusercontent.com/6451685/132952643-cda2d4c5-901b-4f4a-be65-79eb5b8b4d70.JPG)

closes #15447, closes #14064, maybe #14572

glassez commented:
I think there are other suitable candidates to be moved in "Torrent options" dialog. Offhand, these are the following:
1. Save path
2. Sequential download
3. Download first and last piece first

Chocobo1 commented:
>![mid](https://user-images.githubusercontent.com/6451685/132952643-cda2d4c5-901b-4f4a-be65-79eb5b8b4d70.JPG)

I like this better.

thalieht commented:
I've been trying to think about the relationship of autoTMM and save path (should save path be disabled when autoTMM is enabled?) but i now tested how "Set location..." works: it disables autoTMM, so i don't think there should be any interaction between these 2. Let me know if i'm wrong.
Which brings me to the layout.
>I like this better.

Now that i added the save path does that still apply? If yes should i move autoTMM above the Download... checkboxes?
If anyone has a better idea on where to put the new options (autoTMM, save path, Download....) let me know.

![4](https://user-images.githubusercontent.com/6451685/133079625-2ec0ef95-c27b-47c3-b5b4-ef1227c37662.JPG)

>I think there are other suitable candidates to be moved

Moved, as in... remove the originals? Although autoTMM was convenient for me to have it in the context menu, moving it was a necessary evil because it was.... accident prone. But save path? Now users would have to do 2 extra clicks for that.

glassez commented:
>Moved, as in... remove the originals? 

Yes. Otherwise what's the point to move it there?
>But save path? Now users would have to do 2 extra clicks for that.

I am always annoyed by these complaints about a couple of extra clicks, unless it is something that is really often used.
>i now tested how "Set location..." works: it disables autoTMM, so i don't think there should be any interaction between these 2. Let me know if i'm wrong.

When I implemented AutoTMM, it could have been one of two: 
1. Explicitly setting the Save Path disables AutoTMM 
2. Setting the Save Path is disabled/prohibited if autoTMM is enabled. 

I chose the first option because of the simpler implementation. But it seems to me more confusing, so now I would prefer the second option (as it is done in the "Add new torrent" dialog).

glassez commented:
>Now that i added the save path does that still apply? If yes should i move autoTMM above the Download... checkboxes?
If anyone has a better idea on where to put the new options (autoTMM, save path, Download....) let me know.

Well, it should at least be located alongside with "Save path".

thalieht commented:
>>Moved, as in... remove the originals?
>
>Yes. Otherwise what's the point to move it there?

Consolidation, for when people want to change e.g. 3 things at the same time they don't have to open 3 different dialogs.

>I am always annoyed by these complaints about a couple of extra clicks, unless it is something that is really often used.

You're right, for options like the ones in the advanced tab, which you set and forget. Hide them behind 5 clicks, no problem. But you can't call a "volatile" option such as the save path that some people may want to change for every torrent as "not often used". Anyway i don't really care about it personally. You want the shortcuts gone, they're gone.

glassez commented:
>You want the shortcuts gone, they're gone.

That was just my opinion. Other members may still worry about a couple of extra clicks.

Chocobo1 commented on src/gui/torrentoptionsdialog.ui:
```diff
@@ -6,14 +6,44 @@
+   <item>
+    <widget class="QCheckBox" name="checkAutoTMM">
+     <property name="toolTip">
+      <string>Automatic mode means that various torrent properties(e.g. save path) will be decided by the associated category</string>
```
```suggestion
      <string>Automatic mode means that various torrent properties (e.g. save path) will be decided by the associated category</string>
```

Chocobo1 commented on src/gui/torrentoptionsdialog.h:
```diff
@@ -74,10 +76,14 @@ private slots:
     {
         qreal ratio;
         int seedingTime;
+        Qt::CheckState autoTMM;
```
I would prefer they are in this order:
```c++
QString savePath;
qreal ratio;
int seedingTime;
int upSpeedLimit;
int downSpeedLimit;
Qt::CheckState autoTMM;
Qt::CheckState disableDHT;
//...
```

Chocobo1 commented on src/gui/torrentoptionsdialog.cpp:
```diff
@@ -76,11 +84,28 @@ TorrentOptionsDialog::TorrentOptionsDialog(QWidget *parent, const QVector<BitTor
     for (const BitTorrent::Torrent *torrent : torrents)
     {
         m_torrentIDs << torrent->id();
+        if (allSameAutoTMM)
```
maybe add an empty line above.

Chocobo1 commented on src/gui/torrentoptionsdialog.ui:
```diff
@@ -6,14 +6,44 @@
+    <layout class="QHBoxLayout" name="horizontalLayout">
+     <item>
+      <widget class="QLabel" name="labelSavePath">
+       <property name="sizePolicy">
```
This should be reset to default.

Chocobo1 commented on src/gui/torrentoptionsdialog.ui:
```diff
@@ -6,14 +6,44 @@
+      </widget>
+     </item>
+     <item>
+      <widget class="FileSystemPathLineEdit" name="savePath" native="true"/>
```
And set "Horizontal policy" to "Expanding".

thalieht replied:
All comments addressed so far.

Chocobo1 commented:
>But you can't call a "volatile" option such as the save path that some people may want to change for every torrent as "not often used". Anyway i don't really care about it personally. You want the shortcuts gone, they're gone.

No strong opinion from me.

jagannatharjun commented on src/gui/torrentoptionsdialog.cpp:
```diff
@@ -76,11 +84,28 @@ TorrentOptionsDialog::TorrentOptionsDialog(QWidget *parent, const QVector<BitTor
         m_torrentIDs << torrent->id();
+        if (allSameAutoTMM)
+        {
+            if (torrent->isAutoTMMEnabled() != isFirstTorrentAutoTMMEnabled)
```
can't we simplify this code with something like 

```diff
diff --git a/src/gui/torrentoptionsdialog.cpp b/src/gui/torrentoptionsdialog.cpp
index 49ac52cf7..09eddc0af 100644
--- a/src/gui/torrentoptionsdialog.cpp
+++ b/src/gui/torrentoptionsdialog.cpp
@@ -53,6 +53,24 @@ namespace
             slider->setMaximum(value);
         slider->setValue(value);
     }
+
+    template<typename C, typename F>
+    bool allSame(const C &container, const F &unary)
+    {
+        auto b = std::begin(container);
+        const auto e = std::end(container);
+
+        if (b == e) return true;
+
+        const auto v = std::invoke(unary, *b++);
+        while (b != e)
+        {
+            if (v != std::invoke(unary, *b++))
+                return false;
+        }
+
+        return true;
+    }
 }

 TorrentOptionsDialog::TorrentOptionsDialog(QWidget *parent, const QVector<BitTorrent::Torrent *> &torrents)
@@ -64,7 +82,7 @@ TorrentOptionsDialog::TorrentOptionsDialog(QWidget *parent, const QVector<BitTor

     Q_ASSERT(!torrents.empty());
     const auto *session = BitTorrent::Session::instance();
-    bool allSameUpLimit = true, allSameDownLimit = true, allSameRatio = true, allSameSeedingTime = true
+    bool allSameUpLimit = allSame(torrents, &BitTorrent::Torrent::downloadLimit), allSameDownLimit = true, allSameRatio = true, allSameSeedingTime = true
             , allTorrentsArePrivate = true, allSameDHT = true, allSamePEX = true,  allSameLSD = true;

     const int firstTorrentUpLimit = qMax(0, torrents[0]->uploadLimit());
```

maybe also see `std::equal`

thalieht replied:
>can't we simplify this code with something like

Maybe it's simplified for those who can read that :). I'm not familiar with std and i wouldn't know how to change/fix something like that (if asked to).

Chocobo1 reviewed (approved):
Code LGTM. Perhaps the commits could be squashed or have some cleanups, not sure how.

thalieht commented:
I squashed all commits.

glassez commented on src/gui/torrentoptionsdialog.cpp:
```diff
@@ -130,8 +156,31 @@ TorrentOptionsDialog::TorrentOptionsDialog(QWidget *parent, const QVector<BitTor
 
+    if (allSameSavePath)
+        m_ui->savePath->setSelectedPath(firstTorrentSavePath);
+    if (m_ui->checkAutoTMM->checkState() != Qt::PartiallyChecked)
```
The following will look more consistent:

```suggestion
    if (allSameAutoTMM)
```

I am also sure that you should not change state of `m_ui->checkAutoTMM` until all the torrents are checked.

glassez replied:
>I am also sure that you should not change state of `m_ui->checkAutoTMM` until all the torrents are checked.

The same is about other similar checkboxes. You should set these states only after all torrents are traversed:
```c++
if (allSameAutoTMM)
    m_ui->checkAutoTMM->setChecked(isFirstTorrentAutoTMMEnabled);
else
    m_ui->checkAutoTMM->setCheckState(Qt::PartiallyChecked);
```

glassez commented on src/gui/torrentoptionsdialog.cpp:
```diff
@@ -343,6 +414,20 @@ int TorrentOptionsDialog::getSeedingTime() const
+    else
+    {
+        m_ui->labelSavePath->setEnabled(false);
+        m_ui->savePath->setEnabled(false);
```
Shouldn't you also show automatically assigned save path here? Don't remember how "Add new torrent" dialog does it...

thalieht replied:
All done.

glassez reviewed (commented):
Shouldn't the Categories also be moved here to have all the settings associated with AutoTMM in one place?

thalieht commented:
>Shouldn't the Categories also be moved here to have all the settings associated with AutoTMM in one place?

That was harder than i thought. I double checked everything but i probably missed something.

ArcticGems commented:
Maybe it would look better if 'Save path' and 'Category' was inside a box just like the 'Torrent speed limits' and 'Torrent share limits' options??

thalieht commented:
>Maybe it would look better if 'Save path' and 'Category' was inside a box just like the 'Torrent speed limits' and 'Torrent share limits' options??

I'm neutral.

xavier2k6 commented:
>Shouldn't the Categories also be moved here to have all the settings associated with AutoTMM in one place?

What about `tags`?

>Maybe it would look better if 'Save path' and 'Category' was inside a box just like the 'Torrent speed limits' and 'Torrent share limits' options??

To me it looks fine from the screenshots.

ArcticGems commented:
> 
> 
> > Maybe it would look better if 'Save path' and 'Category' was inside a box just like the 'Torrent speed limits' and 'Torrent share limits' options??
> 
> I'm neutral.

> 
> > Maybe it would look better if 'Save path' and 'Category' was inside a box just like the 'Torrent speed limits' and 'Torrent share limits' options??
> 
> To me it looks fine from the screenshots.

OK 👍 

IMO by having the autoTMM checkbox above 'Save path' and 'Category', makes it look "unbalanced" and that's why I made the suggestion.

Anyway, it's not important, so keep it as-is. 👍

thalieht commented:
>>Shouldn't the Categories also be moved here to have all the settings associated with AutoTMM in one place?
>
>What about tags?

Beyond my capabilities.

Chocobo1 commented on src/gui/torrentoptionsdialog.ui:
```diff
@@ -6,14 +6,71 @@
-    <width>390</width>
-    <height>450</height>
+    <width>450</width>
+    <height>540</height>
```
please keep the old values.

thalieht replied:
I added the "Download sequential" checkboxes which increased the minimum width and the same goes for height from the other options.

The other comments are addressed.

Chocobo1 commented on src/gui/torrentoptionsdialog.cpp:
```diff
@@ -37,6 +37,7 @@
 #include "base/bittorrent/session.h"
 #include "base/bittorrent/torrent.h"
 #include "base/global.h"
+#include "base/utils/fs.h"
```
wrong order

Chocobo1 commented on src/gui/transferlistwidget.cpp:
```diff
@@ -329,21 +329,6 @@ QVector<BitTorrent::Torrent *> TransferListWidget::getVisibleTorrents() const
-
-    // Actually move storage
-    for (BitTorrent::Torrent *const torrent : torrents)
-        torrent->move(Utils::Fs::expandPathAbs(newLocation));
```
you should remove `#include "base/utils/fs.h"` from this file.

Chocobo1 commented on src/gui/torrentoptionsdialog.cpp:
```diff
@@ -63,9 +64,18 @@ TorrentOptionsDialog::TorrentOptionsDialog(QWidget *parent, const QVector<BitTor
 
     Q_ASSERT(!torrents.empty());
+    m_ui->savePath->setMode(FileSystemPathEdit::Mode::DirectorySave);
+    m_ui->savePath->setDialogCaption(tr("Choose save path"));
```
These 2 lines should move above `Q_ASSERT()`.
Or another way, move `Q_ASSERT()` to the top of this function block.

Chocobo1 commented on src/gui/torrentoptionsdialog.cpp:
```diff
@@ -241,17 +303,38 @@ TorrentOptionsDialog::TorrentOptionsDialog(QWidget *parent, const QVector<BitTor
+    handleTMMChanged();
+
+    connect(m_ui->comboCategory, &QComboBox::currentIndexChanged, this, &TorrentOptionsDialog::handleCategoryChanged);
+    connect(m_ui->checkAutoTMM, &QCheckBox::clicked, this, &TorrentOptionsDialog::handleTMMChanged);
```
preferably these should follow the order on UI, so this line should be atop of the previous line.

Chocobo1 commented on src/gui/torrentoptionsdialog.cpp:
```diff
@@ -343,6 +445,38 @@ int TorrentOptionsDialog::getSeedingTime() const
+
+    if ((m_ui->checkAutoTMM->checkState() == Qt::Checked) && !m_ui->comboCategory->currentText().isEmpty())
+    {
+        QString savePath = BitTorrent::Session::instance()->categorySavePath(m_ui->comboCategory->currentText());
```
`const QString savePath`

Chocobo1 commented on src/gui/torrentoptionsdialog.cpp:
```diff
@@ -343,6 +445,38 @@ int TorrentOptionsDialog::getSeedingTime() const
+        QString savePath;
+        if ((m_ui->checkAutoTMM->checkState() == Qt::Checked) && !m_ui->comboCategory->currentText().isEmpty())
+            savePath = BitTorrent::Session::instance()->categorySavePath(m_ui->comboCategory->currentText());
+        else // mixed or disabled autoTMM or mixed categories
+            savePath = "";
```
you can omit these 2 lines and refactor it further:
```c++
if ((m_ui->checkAutoTMM->checkState() == Qt::Checked) && !m_ui->comboCategory->currentText().isEmpty())
{
    const QString savePath = BitTorrent::Session::instance()->categorySavePath(m_ui->comboCategory->currentText());
    m_ui->savePath->setSelectedPath(Utils::Fs::toNativePath(savePath));
}
```

glassez commented on src/gui/torrentoptionsdialog.cpp:
```diff
@@ -343,6 +445,37 @@ int TorrentOptionsDialog::getSeedingTime() const
+    {
+        m_ui->labelSavePath->setEnabled(false);
+        m_ui->savePath->setEnabled(false);
+        if ((m_ui->checkAutoTMM->checkState() == Qt::Checked) && !m_ui->comboCategory->currentText().isEmpty())
```
Why is checking for `!m_ui->comboCategory->currentText().isEmpty()` needed?

thalieht replied:
Because it will show the default save path if it's empty.

glassez replied:
> Because it would show the default save path if it's empty.

And what do you expect to see in this case?
"Auto managed" torrents with "empty" category are saved under default save path, aren't they?

thalieht replied:
Empty category also means mixed categories (in this dialog) so i missed that case. I'll pass `allSameCategory` to `handleTMMChanged()` and use that instead.

thalieht replied:
> I'll pass `allSameCategory` to `handleTMMChanged()` and use that instead.

Doesn't seem to work with connect(). Should i make `allSameCategory` member var or maybe there is another way?

glassez replied:
> Should i make allSameCategory member var or maybe there is another way?

Why not?
But I would make it in a slightly different way. I would declare `m_commonCategory` member as `std::optional<QString>` and use it for both common category name and indicator of its presence. Do you understand what I mean?

glassez replied:
On the other hand, it's not a problem to get the name of a common category, just knowing that it exists (we still have a list of all torrents), so a Boolean member variable is enough.

Offtopic: I just generally dislike the existing names like `allSameXXX`.

thalieht replied:
>Do you understand what I mean?

Yep, now that i've read up a little on std::optional. I think the bool is simpler though so i used that.

>Offtopic: I just generally dislike the existing names like allSameXXX.

I'm open to suggestions... that don't involve refactoring.

glassez commented on src/gui/torrentoptionsdialog.cpp:
```diff
@@ -76,11 +85,29 @@ TorrentOptionsDialog::TorrentOptionsDialog(QWidget *parent, const QVector<BitTor
     const bool isFirstTorrentDHTDisabled = torrents[0]->isDHTDisabled();
     const bool isFirstTorrentPEXDisabled = torrents[0]->isPEXDisabled();
     const bool isFirstTorrentLSDDisabled = torrents[0]->isLSDDisabled();
+    const bool isFirstTorrentLSequentialEnabled = torrents[0]->isSequentialDownload();
```
Typo: isFirstTorrent**L**SequentialEnabled
```suggestion
    const bool isFirstTorrentSequentialEnabled = torrents[0]->isSequentialDownload();
```

thalieht replied:
Fixed

Chocobo1 commented:
@thalieht 
The CI errors are strange, try do a `git rebase master`.

thalieht commented:
I had to re-add `#include "base/utils/fs.h"`in transferlistwidget because of https://github.com/qbittorrent/qBittorrent/pull/15468/files#diff-e83414ad4feb1c631ab0dc80a872501d7a6888c87f4e5469fef2451bc6a54dfbR90

glassez commented on src/gui/torrentoptionsdialog.cpp:
```diff
@@ -343,6 +444,37 @@ int TorrentOptionsDialog::getSeedingTime() const
+{
+    Q_UNUSED(index);
+
+    if ((m_ui->checkAutoTMM->checkState() == Qt::Checked) && !m_ui->comboCategory->currentText().isEmpty())
```
```suggestion
    if (m_ui->checkAutoTMM->checkState() == Qt::Checked)
```

glassez reviewed (commented):
IIRC, currently there are no visual differences between "different  categories" and "no category". This is a problem, IMO. An even bigger problem may be if it prevents someone from changing "different  categories" to "no category".

thalieht commented:
>An even bigger problem may be if it prevents someone from changing "different categories" to "no category".

I separated "mixed categories" from "no category" in the category combobox.
/offtopic
You guys probably already know this but i worry you're re-reading the whole diff every time:
You can click on "forced-pushed" to see the diff from the last force push.

glassez commented:
> > An even bigger problem may be if it prevents someone from changing "different categories" to "no category".
> 
> I separated "mixed categories" from "no category" in the category combobox. EDIT: Changed signal for category combobox from `QComboBox::currentIndexChanged` to `QComboBox::activated` because if the "Currently used categories" item was already selected and it was clicked again, it would fill the combobox with the text that should be placeholder.

I'm sorry, I need more time to sort this out.
But first of all I would like to clarify the intended behavior. I am interested in whether we should allow some setting to return to its original state if it was "mixed"? For example, should we allow the user to set the checkboxes to a "grayed" state after it has left it once? Or the category back to "Currently used categories"?

thalieht commented:
> I am interested in whether we should allow some setting to return to its original state if it was "mixed"?

Yes IMO.

glassez commented:
> > I am interested in whether we should allow some setting to return to its original state if it was "mixed"?
> 
> Yes IMO.

But IIRC this is not implemented currently even for simpler properties! Why break your brain trying to do this in such a difficult case, if it will also generate inconsistency?

thalieht commented:
>Well, then it should be available only if the checkbox was Qt::PartiallyChecked initially

Of course, and that's how it is.
>What controls do you mean exactly? Maybe they can have some "special" or empty value?

Speed and share limits. Speed limits use the special value as "inequality" for mixed torrents and if the limit is changed, then it's used for "no limits", and share limits have no radiobutton checked for mixed. I can't find a way to allow users to uncheck a radiobutton (click on the currently checked to uncheck it so all radiobuttons are unchecked).

glassez commented:
> > I can't find a way to allow users to uncheck a radiobutton (click on the currently checked to uncheck it so all radiobuttons are unchecked).
> 
> IIRC, you can play with [QButtonGroup](https://doc.qt.io/qt-5/qbuttongroup.HTML). It allows you to have radio buttons unchecked when its "exclusive" property is set to `false`. Unfortunately you can't keep it always `false` since it allows you to check more than one radiobutton. But you can set it to `false` once already checked radio buttons is clicked and set it to `true` once unchecked radiobutton is clicked.

glassez commented on src/gui/torrentoptionsdialog.cpp:
```diff
@@ -343,8 +454,79 @@ int TorrentOptionsDialog::getSeedingTime() const
+
+void TorrentOptionsDialog::handleTMMChanged()
+{
+    if (!m_ui->checkAutoTMM->isChecked())
```
Anyway you should use `checkState()` here to avoid confusion.

thalieht replied:
>Does isChecked() return true in case check box is partially checked?

Yes, i only found out about it in this PR, which is why i started using `checkState()`. Everything i've done in this dialog was under the assumption that `isChecked()` is true only for checked. Fortunately it doesn't seem to matter in other cases (by chance).

>Anyway you should use checkState() here to avoid confusion.

Done.

glassez reviewed (commented):
Most likely, the latest comments are from me. There are still some things here that cause me complaints, but I think it's better to merge it earlier so that it's easier to understand them. In addition, it overlaps with my current work on improving folder management for incomplete torrents, so it should be merged so that I can continue.

thalieht commented on src/gui/torrentoptionsdialog.cpp:
```diff
@@ -343,8 +454,79 @@ int TorrentOptionsDialog::getSeedingTime() const
+            m_ui->savePath->setSelectedPath(Utils::Fs::toNativePath(savePath));
+        }
+    }
+    else if (m_ui->checkAutoTMM->checkState() == Qt::PartiallyChecked)
```
>If AutoTMM is mixed then "save path" should be already empty and disabled, shouldn't it?

Yes but consider the case of mixed autoTMM and same category. I had to do this somewhere but apparently this is the wrong place. Moved it to `handleTMMChanged()`

glassez replied:
> but consider the case of mixed autoTMM and same category

Generally it doesn't mean that all the torrents have same save path.

Chocobo1 commented on src/gui/torrentoptionsdialog.cpp:
```diff
@@ -109,27 +137,61 @@ TorrentOptionsDialog::TorrentOptionsDialog(QWidget *parent, const QVector<BitTor
+        m_ui->comboCategory->setCurrentText(firstTorrentCategory);
+        m_ui->comboCategory->addItem(firstTorrentCategory);
+    }
+    m_ui->comboCategory->addItem("");
```
This is an empty item, is it correct?
If it is intended, then try `m_ui->comboCategory->addItem(QString());`

thalieht replied:
All comments addressed.
>This is an empty item, is it correct?

Yes, for "no category".

Chocobo1 commented on src/gui/torrentoptionsdialog.cpp:
```diff
@@ -343,8 +454,79 @@ int TorrentOptionsDialog::getSeedingTime() const
+    }
+    else
+    {
+        m_ui->comboCategory->lineEdit()->setPlaceholderText(QLatin1String(""));
```
try `m_ui->comboCategory->lineEdit()->setPlaceholderText({});`
or
``m_ui->comboCategory->lineEdit()->setPlaceholderText(QString());``

Chocobo1 commented on src/gui/torrentoptionsdialog.cpp:
```diff
@@ -343,8 +454,79 @@ int TorrentOptionsDialog::getSeedingTime() const
+    {
+        if (!m_allSameCategory && (m_ui->comboCategory->currentIndex() == 0))
+        {
+            m_ui->savePath->setSelectedPath(QLatin1String(""));
```
try `m_ui->savePath->setSelectedPath({});`
or
`m_ui->savePath->setSelectedPath(QLatin1String(""));`

Chocobo1 commented on src/gui/torrentoptionsdialog.cpp:
```diff
@@ -343,8 +454,79 @@ int TorrentOptionsDialog::getSeedingTime() const
+        {
+            if (!m_allSameCategory && (m_ui->comboCategory->currentIndex() == 0))
+            {
+                m_ui->savePath->setSelectedPath(QLatin1String(""));
```
ditto

Chocobo1 commented on src/gui/torrentoptionsdialog.cpp:
```diff
@@ -343,8 +454,79 @@ int TorrentOptionsDialog::getSeedingTime() const
+        }
+        else // partially checked
+        {
+            m_ui->savePath->setSelectedPath(QLatin1String(""));
```
ditto

Chocobo1 commented on src/gui/torrentoptionsdialog.cpp:
```diff
@@ -343,8 +454,79 @@ int TorrentOptionsDialog::getSeedingTime() const
+    if ((m_initialValues.ratio == MIXED_SHARE_LIMITS) || (m_initialValues.seedingTime == MIXED_SHARE_LIMITS))
+    {
+        QAbstractButton *currentRadio = m_ui->buttonGroup->checkedButton();
+        if ((currentRadio == m_previousRadio) && currentRadio)
```
I would swap the predicates (check for nullptr first): `if (currentRadio && (currentRadio == m_previousRadio))`

Chocobo1 commented on src/gui/torrentoptionsdialog.cpp:
```diff
@@ -343,8 +454,79 @@ int TorrentOptionsDialog::getSeedingTime() const
+        {
+            m_ui->buttonGroup->setExclusive(false);
+            currentRadio->setChecked(false);
+            m_ui->buttonGroup->setExclusive(true);
```
This is unusual, please add some comment within this block about why this is required.

Chocobo1 commented:
@thalieht 
Thank you!

debiedowner commented:
Since @glassez mentioned his annoyance by the "complaints about a couple of extra clicks, unless it is something that is really often used", let me discuss why I used to use the "Download in sequential order" & "Download first and last pieces first" options in the context menu very often. Of course, while I think that my use case is reasonable, I don't know how common it is, but let me explain it anyway as a data point.

When adding a torrent of a media file, I usually select "Download in sequential order" in the Add Torrent dialog, in order to be able to preview the media file to make sure it's correct, ensure its quality etc. Once enough of the file is downloaded and I previewed it, I turn off the "Download in sequential order" option to let it download most optimally for the swarm. Occasionally, I find that the media file also needs the end of the file to open, in which case I select the "Download first and last pieces first" option. I don't select it in the Add Torrent dialog for all torrents, because often it is not needed, and it can take a long time in slow torrents and delay being able to preview the file, especially since #3816 significantly enlarged the percentage to be downloaded at the end and there is no option to decrease it.

So I used to use these two options at least once for *almost every single torrent*, often twice, and sometimes more. This change, which makes it more cumbersome to change these options, encourages me to leave the torrents at downloading sequentially. #164 has a lengthy discussion of why that might be bad. So I was disappointed to see that qBittorrent 4.4.0 pushes me to bad behavior for the health of the swarm, in addition to make things more inconvenient. Also, I am less likely to notice that I left the file downloading sequentially and I should change it, since now this information is hidden in the Torrent options dialog, whereas it used to be always in-my-face as it appeared as a check mark in the context menu.

In any case, I don't expect this change to be taken back, and if any options were to come back to the context menu, I guess those are more likely to be the Category option or maybe Set Location action, as they seem to be getting more complaints (#15946, #15965). Since the other two are unlikely two come back, I would at least love to be able to open the Torrent options more quickly, with a hotkey, as I mentioned in this issue: #15980

glassez commented:
> Since @glassez mentioned his annoyance by the "complaints about a couple of extra clicks, unless it is something that is really often used"

Please don't take my words out of context. It really annoys me when it comes to things that are not so often used.


> let me discuss why I used to use the "Download in sequential order" & "Download first and last pieces first" options in the context menu very often

It is a pity that you (and everyone else who is affected by these changes) did not participate in this topic when it was opened. Well, I get your point. Apparently, we need to return these items to the context menu. Ideally, I would suggest making this menu configurable.

debiedowner commented:
My apologies, re-reading what I wrote, I realized that quoting you in that fashion can be read like I am disagreeing with you, and even being confrontational, which was not at all what I intended. Of course you are right that complaining about extra clicks are pointless without making a case that they need to be used often. I started with that quote merely to give a justification for barging in to the discussion that is already finished; maybe if I had said "Since @glassez rightfully mentioned his annoyance..." it would be more in the tone I intended it, or better yet, I should have omitted it altogether.

And you are also right that I should have made this point earlier before this change was made. I recognize that I am late, that's why I didn't want to be one of those "change was bad, reverse it" people creating new issues and complaining. I had actually made peace with the fact that the decision was made to move them to the Torrent options dialog, so I was just proposing that a way to access that dialog more easily could soften the blow, as @thalieht had earlier discussed. I explained my reasons for using these two options often only to support that proposal, though I would of course be happy to see them return to the context menu.
